### ISSUE TITLE
Investigate ESM support in bokehjs' build

### ISSUE THREAD

**Issue by mattpap** [bokeh/bokeh#8801]
Based on discussion in PR #8782.

> `TypeError: FlatBush is not a constructor`

The underlying problem is that flatbush uses an experimental module field in its package.json to define the ES module entry point, along with the standard main field. webpack and similar linkers seem to prefer module over main, despite its non-standard nature. I will need to do proper research on this subject to decide whether this is a good idea for us to support the module field or not (currently I'm leaning towards the later, due to non-standard by default nature). One can fix this issue on the client side by setting up the linker (assuming webpack) with:
```js
{ resolve: { mainFields: ['main', 'module'] } }
```
For rollup etc. similar configuration option should be hopefully possible.

**Comment by davegravy** [comment]
I ran into this issue.

FWIW I use create-react-app and my (limited) understanding is that modifying webpack config isn't possible without 'ejecting' the project (which I want to avoid for now). If there's another solution I'm all ears.

**Comment by foobarbecue** [comment]
Just ran into this issue and am using Parcel... hopefully I can figure out a workaround.

**Comment by fbove** [comment]
I've tried webpack mainFields configuration but the issue prevails.

Using it this way:
```
import * as Bokeh from 'bokehjs'
// ...
Bokeh.embed.embed_item(data_from_backend)
```
Gives the following error:
```
ERROR TypeError: FlatBush is not a constructor at new SpatialIndex
```
Doing some debugging..
In `node_modules/bokehjs/build/js/lib/core/util/spatial.js`, changing `this.index = new FlatBush(points.length);` to `this.index = new FlatBush.default(points.length);` makes Bokehjs works.

So I guess it is an export module issue.

Is there some way to workaround this?

**Comment by mattpap** [comment]
> `new FlatBush.default`

This most likely means that ES module is still being picked up instead of the commonjs (or UMD in this case) one. flatbush uses different styles of exports depending on the type of the module system.

**Comment by bryevdv** [comment]
Is this something we would help improve by submitting a PR upstream?

**Comment by mattpap** [comment]
I would need to get a reproducible example to decide if this is an issue or not.

**Comment by fbove** [comment]
> I would need to get a reproducible example to decide if this is an issue or not.

A clean webpack project that use bokehjs with sample data from Bokeh docs:
[https://github.com/fbove/bokehjs-example](https://github.com/fbove/bokehjs-example)

Hope it helps!

**Comment by bryevdv** [comment]
Just noting that I ran into this today trying to help @shughes-uk use Bokeh and this will probably be a showstopper for that work. Now that `json_item` and `Bokeh.embed.embed_items` it's much more important to get this use case of bundling BokehJS in a SPA working.

**Comment by bryevdv** [comment]
@mattpap is there a reason we can't just patch this in the package we publish to npmjs.org? I can confirm that the small change above gets the example repo above working. Could we publish a 1.4.1  version on npmjs with this fix?

**Comment by mattpap** [comment]
That will break bokehjs in nodejs, if that matters at all. If you want to cover this case, then you can use conditional code that checks if `default` export exists.

**Comment by mattpap** [comment]
In any case, PR #9348 will fix this properly in 2.0.

**Comment by bryevdv** [comment]
> . If you want to cover this case, then you can use conditional code that checks if default export exists.

That seems ideal, uploading a 1.4.1 version patched this way would leave a BokehJS that could work with anyone stuck on 1.4 for Python 2 reasons (and also let us create and publicize small sample react/angular/etc examples very soon)

> In any case, PR #9348 will fix this properly in 2.0.

For sure, it seems like a tiny effort for a fairly important immediate gain. Can you create the patched packages? I can handle publishing them if you can get them to me.

**Comment by gsaandy** [comment]
> Based on discussion in PR #8782.
> 
> > `TypeError: FlatBush is not a constructor`
> 
> The underlying problem is that flatbush uses an experimental module field in its package.json to define the ES module entry point, along with the standard main field. webpack and similar linkers seem to prefer module over main, despite its non-standard nature. I will need to do proper research on this subject to decide whether this is a good idea for us to support the module field or not (currently I'm leaning towards the later, due to non-standard by default nature). One can fix this issue on the client side by setting up the linker (assuming webpack) with:
> 
> ```js
> { resolve: { mainFields: ['main', 'module'] } }
> ```
> 
> For rollup etc. similar configuration option should be hopefully possible.

There are chances that this could break other modules

It worked for me after using webpack alias
```
alias: {
  flatbush: 'flatbush/flatbush'
}
```
